File: frontend/src/contracts/states.ts
export const SESSION_STATES = ["CREATED", "ACTIVE", "REVIEW", "FINALIZED"] as const;
export type SessionState = (typeof SESSION_STATES)[number];

export const COUNT_LINE_STATES = [
  "DRAFT",
  "SUBMITTED",
  "PENDING_APPROVAL",
  "APPROVED",
  "REJECTED",
  "LOCKED",
] as const;
export type CountLineState = (typeof COUNT_LINE_STATES)[number];

const SESSION_ALIASES: Record<string, SessionState> = {
  OPEN: "CREATED",
  CREATED: "CREATED",
  ACTIVE: "ACTIVE",
  PAUSED: "ACTIVE",
  REVIEW: "REVIEW",
  RECONCILE: "REVIEW",
  RECONCILING: "REVIEW",
  COMPLETED: "FINALIZED",
  CLOSED: "FINALIZED",
  FINALIZED: "FINALIZED",
  CANCELLED: "FINALIZED",
};

export function normalizeSessionState(value: unknown): SessionState | "UNKNOWN" {
  const raw = String(value ?? "").trim().toUpperCase();
  if (!raw) return "UNKNOWN";
  return SESSION_ALIASES[raw] || "UNKNOWN";
}

const COUNT_LINE_ALIASES: Record<string, CountLineState> = {
  DRAFT: "DRAFT",
  SUBMITTED: "SUBMITTED",
  PENDING_APPROVAL: "PENDING_APPROVAL",
  PENDING: "PENDING_APPROVAL",
  APPROVED: "APPROVED",
  REJECTED: "REJECTED",
  LOCKED: "LOCKED",
  FINALIZED: "LOCKED",
};

export function normalizeCountLineState(value: unknown): CountLineState | "UNKNOWN" {
  const raw = String(value ?? "").trim().toUpperCase();
  if (!raw) return "UNKNOWN";
  return COUNT_LINE_ALIASES[raw] || "UNKNOWN";
}

File: backend/contracts/states.py
"""
Canonical state contracts shared across backend modules.
"""

from __future__ import annotations

from typing import Final

SESSION_STATES: Final[tuple[str, ...]] = (
    "CREATED",
    "ACTIVE",
    "REVIEW",
    "FINALIZED",
)

COUNT_LINE_STATES: Final[tuple[str, ...]] = (
    "DRAFT",
    "SUBMITTED",
    "PENDING_APPROVAL",
    "APPROVED",
    "REJECTED",
    "LOCKED",
)

LEGACY_SESSION_STATUS_ALIASES: Final[dict[str, str]] = {
    "OPEN": "CREATED",
    "ACTIVE": "ACTIVE",
    "PAUSED": "ACTIVE",
    "RECONCILE": "REVIEW",
    "RECONCILING": "REVIEW",
    "REVIEW": "REVIEW",
    "COMPLETED": "FINALIZED",
    "CLOSED": "FINALIZED",
    "CANCELLED": "FINALIZED",
    "FINALIZED": "FINALIZED",
    "CREATED": "CREATED",
}

COUNT_LINE_STATUS_ALIASES: Final[dict[str, str]] = {
    "FINALIZED": "LOCKED",
    "LOCKED": "LOCKED",
    "APPROVED": "APPROVED",
    "REJECTED": "REJECTED",
    "PENDING": "PENDING_APPROVAL",
    "PENDING_APPROVAL": "PENDING_APPROVAL",
    "SUBMITTED": "SUBMITTED",
    "DRAFT": "DRAFT",
}


def normalize_session_state(value: object) -> str:
    raw = str(value or "").strip().upper()
    if not raw:
        return "UNKNOWN"
    return LEGACY_SESSION_STATUS_ALIASES.get(raw, raw)


def normalize_count_line_state(value: object) -> str:
    raw = str(value or "").strip().upper()
    if not raw:
        return "UNKNOWN"
    return COUNT_LINE_STATUS_ALIASES.get(raw, raw)
